### ISSUE TITLE
TS2394: This overload signature is not compatible with its implementation signature.

### ISSUE THREAD

**Issue by NarHakobyan** [microsoft/TypeScript#57596]
### 🔎 Search Terms

TS2394, TS2394: This overload signature is not compatible with its implementation signature.

### 🕗 Version & Regression Information

- This is a crash
- This changed between version 5.3.3

### 💻 Code

```ts
  async getOne(
    id: Uuid,
    options?: { language?: LanguageCodeEnum; indexed?: false },
  ): Promise<string>;

  async getOne(
    id?: Uuid | null,
    options?: { language?: LanguageCodeEnum; indexed?: false },
  ): Promise<string | undefined>;

  async getOne(
    id: Uuid,
    options?: { language?: LanguageCodeEnum; indexed: true },
  ): Promise<number>;

  async getOne(
    id?: Uuid | null,
    options?: { language?: LanguageCodeEnum; indexed: true },
  ): Promise<number | undefined>;

  async getOne(
    id?: Uuid | null,
    options?: { language?: LanguageCodeEnum; indexed?: boolean },
  ): Promise<string | number | undefined> {
    // to prevent check every time in the code

    return await options as any;
  }
```

### 🙁 Actual behavior

always gives an error on first overload 

TS2394: This overload signature is not compatible with its implementation signature.

### 🙂 Expected behavior

error must be gone

**Comment by NarHakobyan** [comment]
![Uploading image.png…]()

**Comment by snarbies** [comment]
This is all I see in the error pane in the [playground](https://www.typescriptlang.org/play?#code/KYOwrgtgBAMghiA5mOjgGED2ATYBRcCAbwF8AoAFwE8AHYKAVTAEtsoBeKIgZ0wmABcUAOQUAFsyTDyZAMYAbON25Q8ADzgQa84ETJkoUJVRCyoaCgHkQwABQHDUVkKasANA8OYaFZphDcAPxCRFCKSChowbAIyKgYOPiEANxOILhqwNjRAGZw8tz0JB6GAJRCAAoATnzMhQA83BRVkogAfMn6hsam5sBWNvaOTtkuLGwAPlDg8vIljt6+-kEhYbGRwNHwEfFYuASQqZIZWbn5hVDFDuVQ1bUNTS1IUFNg6cA5klkdXUbcJmYLNY7J4RmN3KDFn4AtFQuE4lEhNsEQl9ik0idsEJmmAivMbncIHVgPVCAAjYBVH4OHqA-rAoaOVjRVyTaZgWbzLw+aErLhrHaImKC1FJQ4Y4CZLFQHF466VGpEhrkykvKBvXCfGzYandf69IGDUHM8FsmZzSE85awgUorbrXaJA4QI7vKXRMmYTA6BCXfEK+4kx6tNUqqpqjUfL46riggD0cZlmCgNCqwAAbqAKFBZGJgLIANZQDOUqgy5j8NIyvM5xK-QxpihgKogIwAdzgzGzUOWfyMIConUMMhIQA):

```
Errors in code

    'id' is declared but its value is never read.
    'await' has no effect on the type of this expression.
```

**Comment by jcalz** [comment]
> * This is a crash

No, a crash is when `tsc` itself fails to run. (Maybe the template should say this?)

 
> * This changed between version 5.3.3

This doesn't make much sense. Maybe you meant to use "This is the behavior in every version I tried"?

> ### ⏯ Playground Link
> 
> _No response_

This doesn't help, please put a `https://www.typescriptlang.org/play/...` link to a reproduction of your issue. If we can't reproduce it then we can't do much about it.

**Comment by NarHakobyan** [comment]
this happens when I add extra function on promise

```
declare global {
  //   add try function on Promise
  interface Promise<T> {
    try: (this: Promise<T>) => Promise<T | null>;
  }
}
`
